store/commitment/iavl/tree.go

```diff
@@ -83,6 +94,142 @@ func (t *IavlTree) Prune(version uint64) error {
 	return t.tree.DeleteVersionsTo(int64(version))
 }
 
+// Snapshot implements snapshottypes.CommitSnapshotter.
+func (t *IavlTree) Snapshot(version uint64, protoWriter protoio.Writer) error {
+	if version == 0 {
+		return fmt.Errorf("the snapshot version must be greater than 0")
+	}
+
+	latestVersion := t.GetLatestVersion()
+	if version > latestVersion {
+		return fmt.Errorf("the snapshot version %d is greater than the latest version %d", version, latestVersion)
+	}
+
+	tree, err := t.tree.GetImmutable(int64(version))
+	if err != nil {
+		return fmt.Errorf("failed to get immutable tree for version %d: %w", version, err)
+	}
+
+	exporter, err := tree.Export()
+	if err != nil {
+		return fmt.Errorf("failed to export tree for version %d: %w", version, err)
+	}
+
+	defer exporter.Close()
+
+	err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+		Item: &snapshottypes.SnapshotItem_Store{
+			Store: &snapshottypes.SnapshotStoreItem{
+				Name: t.storeKey,
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to write store name: %w", err)
+	}
+
+	for {
+		node, err := exporter.Next()
+		if errors.Is(err, iavl.ErrorExportDone) {
+			break
+		} else if err != nil {
+			return fmt.Errorf("failed to get the next export node: %w", err)
+		}
+
+		if err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+			Item: &snapshottypes.SnapshotItem_IAVL{
+				IAVL: &snapshottypes.SnapshotIAVLItem{
+					Key:     node.Key,
+					Value:   node.Value,
+					Height:  int32(node.Height),
+					Version: node.Version,
+				},
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to write iavl node: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Restore implements snapshottypes.CommitSnapshotter.
+func (t *IavlTree) Restore(version uint64, format uint32, protoReader protoio.Reader, chStorage chan<- *store.KVPair) (snapshottypes.SnapshotItem, error) {
+	var (
+		importer     *iavl.Importer
+		snapshotItem snapshottypes.SnapshotItem
+	)
+
+loop:
+	for {
+		snapshotItem = snapshottypes.SnapshotItem{}
+		err := protoReader.ReadMsg(&snapshotItem)
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return snapshottypes.SnapshotItem{}, fmt.Errorf("invalid protobuf message: %w", err)
+		}
+
+		switch item := snapshotItem.Item.(type) {
+		case *snapshottypes.SnapshotItem_Store:
+			t.storeKey = item.Store.Name
+			importer, err = t.tree.Import(int64(version))
+			if err != nil {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("failed to import tree for version %d: %w", version, err)
+			}
+			defer importer.Close()
+
+		case *snapshottypes.SnapshotItem_IAVL:
+			if importer == nil {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("received IAVL node item before store item")
+			}
+			if item.IAVL.Height > int32(math.MaxInt8) {
```

Comment:
You can make the code simple and avoid long repetitions in the error by
```go
if g, w := item.IAVL.Height, int32(math.MaxInt8); g > w {
     return si, fmt.Errorf("node height %d must not exceed %d", g, w)
}
```

after having declared your function signature as
```go
func (t *IavlTree) Restore(version uint64, format uint32, protoReader protoio.Reader, chStorage chan<- *store.KVPair) (si snapshottypes.SnapshotItem, err error) {
```